lib/rules/string-no-newline/index.mjs

```diff
@@ -56,67 +55,54 @@ const rule = (primary) => {
 				return;
 			}
 
-			parseSelector(ruleNode.selector, result, ruleNode)?.walkAttributes((attributeNode) => {
-				const { value, quoteMark } = attributeNode;
-
-				if (!value || !reNewLine.test(value)) {
-					return;
-				}
-
-				const openIndex = [
-					// length of our attribute
-					attributeNode.attribute,
-					// length of our operator , ie '='
-					attributeNode.operator || '',
-				].reduce(
-					(index, str) => index + str.length,
-					// index of the start of our attribute node in our source
-					// plus 1 for the opening quotation mark
-					attributeNode.sourceIndex + 1,
-				);
-
-				const valueLength = value.length + (quoteMark || '').length * 2;
-
-				report({
-					message: messages.rejected,
-					node: ruleNode,
-					index: openIndex,
-					endIndex: openIndex + valueLength,
-					result,
-					ruleName,
-				});
-			});
+			check(ruleNode, ruleNode.selector, () => 0);
 		}
 
 		/**
-		 * @template {import('postcss').AtRule | import('postcss').Declaration} T
+		 * @template {import('postcss').AtRule | import('postcss').Rule | import('postcss').Declaration} T
 		 * @param {T} node
 		 * @param {string} value
 		 * @param {(node: T) => number} getIndex
 		 * @returns {void}
 		 */
-		function checkDeclOrAtRule(node, value, getIndex) {
+		function check(node, value, getIndex) {
 			// Get out quickly if there are no new line
 			if (!reNewLine.test(value)) {
 				return;
 			}
 
-			valueParser(value).walk((valueNode) => {
-				if (valueNode.type !== 'string') {
-					return;
-				}
+			/** @typedef {import('@csstools/css-tokenizer').ParseError} ParseError */
+			/** @type {Array<ParseError>} */
+			let errors = [];
+			const tokens = tokenize(
+				{ css: value },
+				{
+					onParseError: (/** @type {ParseError} */ err) => {
+						if (err.message === 'Unexpected newline while consuming a string token.') {
```

Comment:
🤔 It can only be a `ParseError` :

```ts
export function tokenize(
	input: {
		css: { valueOf(): string },
		unicodeRangesAllowed?: boolean
	},
	options?: {
		onParseError?: (error: ParseError) => void
	},
): Array<CSSToken>
```

Checking the string is how we can check that the problem is a newline in a string and not for example a bad url token.

-----

I could provide exported constants for these errors in `@csstools/css-tokenizer`.
That would be easier.

I am not sure what I can do about the type signature so that TypeScript correctly determines that `err` is always of type `ParseError` in this callback.

Reply:
> I could provide exported constants for these errors in `@csstools/css-tokenizer`.
> That would be easier.

That sounds better 👍🏼

Reply:
I've added a new `ParseErrorWithToken` subclass and a `ParseErrorMessage` object holding all the possible error messages.

The `ParseErrorWithToken` (as the name implies) is a parse error that holds a reference to the token that caused the error.

With this we have the error and token in one place without extra loops.

Reply:
Thanks. I love this change. 👍🏼